core/entity.js

```diff
@@ -1,285 +1,57 @@
-import { create_float_buffer, create_index_buffer } from './context';
-import { vec3, mat4, quat } from '../math';
-import { hex_to_rgb } from '../utils';
-
-// When using arrow keys for rotation simulate the mouse delta of this value.
-const KEY_ROTATION_DELTA = 3;
-
 const default_options = {
-  scale: vec3.unit.slice(),
-  position: vec3.zero.slice(),
-  rotation: quat.create(),
-
-  keyboard_controlled: false,
-  mouse_controlled: false,
-
-  move_speed: 3.5,
-  rotate_speed: .5,
-
-  dir_desc: {
-    87: 'f',
-    65: 'l',
-    68: 'r',
-    83: 'b',
-    69: 'u',
-    81: 'd',
-    38: 'pu',
-    40: 'pd',
-    39: 'yr',
-    37: 'yl'
-  },
-
+  components: [],
   skip: false,
 };
 
 export class Entity {
   constructor(options) {
-    this.matrix = mat4.create();
-    this.world_matrix = mat4.create();
-    this.world_to_self = mat4.create();
-
-    this._scale = vec3.unit.slice();
-    this._rotation = quat.create();
-
     Object.assign(this, default_options, options);
 
-    if (this.material) {
-      this.color_vec = [];
-      this.color_opacity = 1.0;
-      this.color = options.color || '#ffffff';
-    }
-
     this.entities = new Set();
+    const components = this.components;
+    this.components = new Map();
```

Comment:
Since you're creating `this.components` as a `new Map()` here anyways, you can remove `components` from `default_options` and just check if `options.components` if present before you run `add_component` them.

Reply:
Actually, let's leave `components` in `default_options` to serve as a contract definition about what the options can be.